Source/WebCore/platform/graphics/CopyImageOptions.h

```diff
@@ -37,4 +37,9 @@ enum class ShouldInterpolate : bool {
     No // Edges of the image may appear pixelated.
 };
 
+enum class VolatileCopy : bool {
+    Yes, // Copy should be destroyed before the subsequent draws.
```

Comment:
So if the caller forgets to destroy the copy, we get a weird bug where the surfaces appear to magically change under us?

Is it possible to implement this using `CFGetRetainCount` to determine if we need copy-on-write behaviour?

Reply:
This is great idea.

And it tried doing it by making ImageBufferIOSurfaceBackend be a NativeImage::Observer. And I made it keep a list of the copied CGImages. And instead of checking whether copyNativeImage() was ever called since the last time the image cache was cleared, I was checking whether this list is not empty. This seems to work except for one this. NativeImage has to be destroyed on the main thread but in GPUP they are copied from the IOSurface on RemoteRenderingBackend work queue.

When running the layout tests I hit many crashes which have the same scenario. On the RemoteRenderingBackend work queue, the IOSurface backend is trying to detach itself from all the copied NativeImages because the backing store was invalidated. And on the main thread the NativeImage is being destroyed and its destructor is trying to detach itself from the IOSurface backend.

I think it is still doable and maybe we want to remove WTF::DestructionThread::Main from the NativeImage. But I think this should be addressed in a separate patch.

Reply:
Ideally, "Yes" wouldn't come first in the enum. It is weird for Yes to have 0 as value. That's why we usually use { No, Yes } order in such enum classes.